### PR TITLE
[PiranhaJava] Allow for custom test annotations to be specified in .json configuration

### PR DESCRIPTION
**Pull request by lazaroclapp**
Initial implementation of #115 (without support for groups)

We add a new sub-package to abstract away handling of test annotations and matching of actual annotation instances in code to their best matching spec in the provided configuration file. See #115 for the general description of test specification language.

Additionally, this PR refactors Piranha configuration into its own package.

Tests included.

### CODE REVIEW

**Comment by lazaroclapp** [comment]
This is a partial WIP (it's missing, at the very least, a couple javadocs). @mkr-plse feel free to do an initial pass for review if you want, but I won't add you as an official reviewer until this is fully done, mostly uploaded it so I could go over this myself.

There is also no support in here yet for specifying treatment groups in test annotations (adding one to a specification in `piranha.json` will result in an exception), but that will left for a follow up PR as this grows quite long as it is. 

We also should really refactor that huge test class into multiple classes (e.g. one to test just annotations related refactoring), but I am leaving it as is in this PR both for size and to make double-sure existing tests pass after this change.

**Comment by mkr-plse** [comment]
> This is a partial WIP (it's missing, at the very least, a couple javadocs). @mkr-plse feel free to do an initial pass for review if you want, but I won't add you as an official reviewer until this is fully done, mostly uploaded it so I could go over this myself.

Please let me know when it is ready for review. I will take a pass then. Thanks!

File: java/piranha/src/main/java/com/uber/piranha/testannotations/TestAnnotationResolver.java [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+      Preconditions.checkArgument(
+          TRUE_STR.equalsIgnoreCase(treatedValueOrField)
+              || FALSE_STR.equalsIgnoreCase(treatedValueOrField));
+      treated = TRUE_STR.equalsIgnoreCase(treatedValueOrField);
+    }
+    // Process treatment group, if any
+    ImmutableList.Builder<AnnotationArgument> groupsBuilder = ImmutableList.builder();
+    Optional<String> groupValueOrField = spec.getGroupValueOrField();
+    if (groupValueOrField.isPresent()) {
+      // ToDo: implement
```
Comment (lazaroclapp) [inline java/piranha/src/main/java/com/uber/piranha/testannotations/TestAnnotationResolver.java:184]: This is intentionally left out of scope for the current PR.

File: java/piranha/src/main/java/com/uber/piranha/testannotations/TestAnnotationResolver.java [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+            NewArrayTree arrayExpression = (NewArrayTree) assnExpression;
+            for (ExpressionTree expr : arrayExpression.getInitializers()) {
+              if (expr instanceof LiteralTree) {
+                builder.put(
+                    key, new AnnotationArgument(((LiteralTree) expr).getValue().toString(), expr));
+              } else {
+                builder.put(
+                    key,
+                    new AnnotationArgument(
+                        ASTHelpers.getSymbol(expr).getSimpleName().toString(), expr));
```
Comment (mkr-plse) [inline java/piranha/src/main/java/com/uber/piranha/testannotations/TestAnnotationResolver.java:99]: Can we have a helper for ASTHelpers.getSymbol(expr).getSimpleName().toString()?

File: java/piranha/src/test/java/com/uber/piranha/XPFlagCleanerTest.java [hunk 3]
```diff
@@ -2525,6 +2526,180 @@ public void testAnnotatedTestMethodsArrayExpressionControl() throws IOException
...
+            "  public void foo () {}",
+            "  @NewToggleTreated(\"STALE_FLAG\")",
+            "  public void x () {}",
+            "  @NewToggleControl(\"STALE_FLAG\")",
+            "  public void y () {}",
+            "  @NewToggleTreated(\"STALE_FLAG\")",
+            "  @NewToggleTreated(\"OTHER_FLAG_1\")",
+            "  public void z () {}",
+            "  @NewToggleControl(\"STALE_FLAG\")",
+            "  @NewToggleTreated(\"OTHER_FLAG_1\")",
```
Comment (mkr-plse) [inline java/piranha/src/test/java/com/uber/piranha/XPFlagCleanerTest.java:2643]: can we also add a function with one annotation for `OTHER_FLAG_1` and ensure that there are no changes?

**Review by mkr-plse — APPROVED** [review]

File: java/piranha/src/main/java/com/uber/piranha/testannotations/TestAnnotationResolver.java (hunk 2 above)
Comment (lazaroclapp) [reply java/piranha/src/main/java/com/uber/piranha/testannotations/TestAnnotationResolver.java:99]: Sure. This is fairly idiomatic EP, but I also don't see the harm in it :)

**Review by mkr-plse — APPROVED** [review]
lgtm.

**Pull request merged**